File: graph traversals 2/checkposts.py
def dfs(graph, v, visited, stack):
    visited[v] = True
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited, stack)
    stack.append(v)

def transpose(graph, n):
    g = [[] for _ in range(n)]
    for i in range(n):
        for j in graph[i]:
            g[j].append(i)
    return g

def fillOrder(graph, n, visited, stack):
    for i in range(n):
        if not visited[i]:
            dfs(graph, i, visited, stack)

def kosaraju(graph, n):
    stack = []
    visited = [False] * n

    fillOrder(graph, n, visited, stack)

    gr = transpose(graph, n)

    visited = [False] * n

    scc_costs = []
    while stack:
        i = stack.pop()
        if not visited[i]:
            scc = []
            dfs(gr, i, visited, scc)
            scc_costs.append(scc)

    return scc_costs

def solve(n, costs, graph):
    MOD = 1000000007

    sccs = kosaraju(graph, n)
    min_cost = 0
    ways = 1

    for scc in sccs:
        min_scc_cost = min(costs[i] for i in scc)
        min_cost += min_scc_cost
        ways = (ways * sum(costs[i] == min_scc_cost for i in scc)) % MOD

    return min_cost, ways 

n = int(input().strip())
costs = list(map(int, input().strip().split()))
m = int(input().strip())

graph = [[] for _ in range(n)]
for _ in range(m):
    u, v = map(int, input().strip().split())
    graph[u - 1].append(v - 1)

min_cost, ways = solve(n, costs, graph)
print(min_cost, ways)

File: graph traversals 2/checkposts.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <climits>
using namespace std;

void dfs(int v, vector<bool>& visited, vector<vector<int>>& graph, vector<int>& result) {
    visited[v] = true;
    for (int i : graph[v]) {
        if (!visited[i]) {
            dfs(i, visited, graph, result);
        }
    }
    result.push_back(v);
}

vector<vector<int>> transpose(const vector<vector<int>>& graph, int n) {
    vector<vector<int>> g(n);
    for (int i = 0; i < n; ++i) {
        for (int j : graph[i]) {
            g[j].push_back(i);
        }
    }
    return g;
}

void fillOrder(vector<vector<int>>& graph, int n, vector<bool>& visited, stack<int>& stack) {
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            vector<int> temp;
            dfs(i, visited, graph, temp);
            for (int node : temp) {
                stack.push(node);
            }
        }
    }
}

vector<vector<int>> kosaraju(vector<vector<int>>& graph, int n) {
    stack<int> stack;
    vector<bool> visited(n, false);
    fillOrder(graph, n, visited, stack);

    vector<vector<int>> gr = transpose(graph, n);
    fill(visited.begin(), visited.end(), false);

    vector<vector<int>> sccs;
    while (!stack.empty()) {
        int i = stack.top();
        stack.pop();
        if (!visited[i]) {
            vector<int> scc;
            dfs(i, visited, gr, scc);
            sccs.push_back(scc);
        }
    }
    return sccs;
}

pair<long long, long long> solve(int n, vector<int>& costs, vector<vector<int>>& graph) {
    const int MOD = 1000000007;
    vector<vector<int>> sccs = kosaraju(graph, n);

    long long min_cost = 0, ways = 1;
    for (const auto& scc : sccs) {
        int min_scc_cost = INT_MAX;
        for (int i : scc) {
            min_scc_cost = min(min_scc_cost, costs[i]);
        }
        min_cost += min_scc_cost;

        int count = 0;
        for (int i : scc) {
            if (costs[i] == min_scc_cost) {
                count++;
            }
        }
        ways = (ways * count) % MOD;
    }
    return {min_cost, ways};
}

int main() {
    int n, m;
    cin >> n;
    vector<int> costs(n);
    for (int& cost : costs) {
        cin >> cost;
    }
    cin >> m;
    vector<vector<int>> graph(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        graph[u - 1].push_back(v - 1);
    }

    auto [min_cost, ways] = solve(n, costs, graph);
    cout << min_cost << " " << ways << endl;
    return 0;
}
